contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,154 @@
+// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
+// OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
+
+// THIS IS NOT SECURITY AUDITED
+// DO NEVER USE THIS WITHOUT SECURITY AUDIT FIRST
+
+@compiler >= 6
+
+include "Option.aes"
+include "Set.aes"
+include "String.aes"
+
+// This contract provides multi signature based on generalized accounts.
+// A transaction can be proposed by one of the signers which are registered during contract deployment.
+// The transaction must be confirmed by a certain amount of signers. This is also defined during contract deployment.
+// Proposing a transaction automatically counts as confirmation.
+// When proposing a tx, the signer can define a TTL until the proposed tx needs to be confirmed and authorized.
+// If the amount of confirmations is reached, the special function "authorize" can be called by anybody.
+// Only one transaction can be handled at the same time.
+// Every signer can revoke the proposed transaction at any time (no matter if it's expired or not).
+contract SimpleGAMultiSig =
+
+  datatype event
+    = TxProposed(hash, address, int)
+    | TxConfirmed(hash, address)
+    | TxConsensusReached(hash)
+    | TxRevoked(hash, address)
+    | TxAuthorized(hash)
+
+  record transaction =
+    { ga_tx_hash : hash,
+      expires_at : int,
+      proposed_by : address,
+      confirmed_by : Set.set(address) }
+
+  record consensus_info =
+    { ga_tx_hash : option(hash),
+      confirmations_required : int,
+      confirmed_by : list(address),
+      has_consensus : bool,
+      expired : bool }
+
+  record state =
+  	{ signers : Set.set(address),
+  	  confirmations_required : int,
+      current_tx : option(transaction) }
+
+  stateful entrypoint init(p_confirmations_required : int, p_cosigners : list(address)) =
+    { signers = Set.insert(Call.caller, Set.from_list(p_cosigners)),
+  	  confirmations_required = p_confirmations_required,
+  	  current_tx = None }
+
+  // GA authorize function
+  stateful entrypoint authorize() : bool =
+    let tx = Option.force_msg(state.current_tx, "ERROR_NO_TX_PROPOSED")
+    require(!is_expired(tx), "ERROR_TX_EXPIRED")
+    switch(Auth.tx_hash)
+      None => abort("ERROR_NO_AUTH_CONTEXT")
+      Some(ga_tx_hash) =>
+        require(has_consensus(tx), "ERROR_NO_CONSENSUS")
+        require(tx.ga_tx_hash == ga_tx_hash, "ERROR_UNEQUAL_HASHES")
+        Chain.event(TxAuthorized(ga_tx_hash))
+        put(state{ current_tx = None })
+        true
+
+  // propose a new tx valid for the given ttl
+  stateful entrypoint propose(ga_tx_hash : hash, ttl : Chain.ttl, sig : signature) =
+    require_valid_signer_and_sig(Call.caller, ga_tx_hash, sig)
+    switch(state.current_tx)
+      None => set_current_tx(ga_tx_hash, ttl)
+      Some(tx) =>
+        require(is_expired(tx), "ERROR_A_TX_IS_ALREADY_PROPOSED")
```

Comment:
shoudn't you rather check `current_tx` here?

Reply:
I think it's correct:

- `state.current_tx` is being accessed with `Some(tx)`

Reply:
either the check or the error message is incorrect then, you check expiration but error for already proposed

Reply:
yeah it might be confusing. it's about proposing a new transaction. and if there is already a tx present it's required to be expired. otherwise it's not allowed to propose a new tx. it needs to be revoked first.

does that explanation make sense?

Reply:
lets adjust the code or error message to fit your explaination (just my opinion, feel free to keep as you think)

Reply:
will change error message to `ERROR_EXISTING_PROPOSED_TX_NOT_EXPIRED`